mlir/lib/Conversion/AIRRtToNpuPass.cpp

```diff
@@ -198,6 +198,65 @@ struct HerdLoadToNpuPattern : public OpConversionPattern<HerdLoadOp> {
   LogicalResult
   matchAndRewrite(HerdLoadOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
+
+    auto module = op->getParentOfType<ModuleOp>();
+
+    // get the size metadata associated with this herd load
+    int64_t size_x = -1;
+    int64_t size_y = -1;
+    int64_t loc_x = -1;
+    int64_t loc_y = -1;
+    module.walk([&](HerdMetadataOp metadata) {
+      // return the first match by name
+      if (metadata.getSymName() != op.getSymName())
+        return WalkResult::advance();
+      auto sxAttr = metadata->getAttrOfType<IntegerAttr>("size_x");
+      auto syAttr = metadata->getAttrOfType<IntegerAttr>("size_y");
+      auto lxAttr = metadata->getAttrOfType<IntegerAttr>("loc_x");
+      auto lyAttr = metadata->getAttrOfType<IntegerAttr>("loc_y");
+      if (sxAttr && syAttr && lxAttr && lyAttr) {
+        size_x = sxAttr.getInt();
+        size_y = syAttr.getInt();
+        loc_x = lxAttr.getInt();
+        loc_y = lyAttr.getInt();
+      } else {
+        metadata.emitWarning(
+            "airrt.herd_metadata missing size_x, size_y, loc_x, or loc_y.");
+      }
+      return WalkResult::interrupt();
+    });
+    if (size_x < 0 || size_y < 0 || loc_x < 0 || loc_y < 0) {
+      op.emitWarning("airrt.herd_metadata missing or incomplete.");
+      return failure();
+    }
+
+    // for each herd core, emit write_rtp ops for every herd operand
+    // followed by a write32 to the herd lock, setting it to 1.
+    for (int phys_x = loc_x; phys_x < size_x + loc_x; phys_x++) {
+      for (int phys_y = loc_y; phys_y < size_y + loc_y; phys_y++) {
+
+        for (int i = 0, e = op.getNumOperands(); i < e; i++) {
+          Value oper = adaptor.getOperands()[i];
+          if (!llvm::isa<IntegerType, IndexType, FloatType>(oper.getType()))
+            continue;
+
+          std::string name = "__air_herd_rtp_" + std::to_string(phys_x) + "_" +
+                             std::to_string(phys_y);
+          auto constOp =
+              dyn_cast_if_present<arith::ConstantOp>(oper.getDefiningOp());
+          if (!constOp)
+            continue;
+          uint32_t v = cast<IntegerAttr>(constOp.getValue()).getInt();
+          rewriter.create<AIEX::NpuWriteRTPOp>(op.getLoc(), name, i, v);
+        }
+        // FIXME: this should depend on the metadata to enable and to get the id
+        if (op.getNumOperands())
+          rewriter.create<AIEX::NpuWrite32Op>(
+              op.getLoc(), 0x0001F000, 0x1, nullptr,
+              rewriter.getI32IntegerAttr(phys_x),
+              rewriter.getI32IntegerAttr(phys_y));
+      }
```

Comment:
This is the code that could be simplified by https://github.com/Xilinx/mlir-aie/pull/2160. The main challenge, however, is tracking which lock was allocated by the air-to-aie pass. The FIXME here refers to the fact that the lock is named (with a symbol name), and that name should be used to lookup which lock to set, instead of assuming lock id=0 (which is true by constuction but not obvious).